flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java

```diff
@@ -0,0 +1,225 @@
+		ChannelStateWriteResult result = new ChannelStateWriteResult();
+		ChannelStateWriteResult put = results.computeIfAbsent(checkpointId, id -> {
+			Preconditions.checkArgument(results.size() < maxCheckpoints, "results.size() > maxCheckpoints", results.size(), maxCheckpoints);
+			enqueue(new CheckpointStartRequest(checkpointId, result, checkpointOptions.getTargetLocation()));
+			return result;
+		});
+		Preconditions.checkArgument(put == result, "result future already present for checkpoint id: " + checkpointId);
+	}
+
+	@Override
+	public void addInputData(long checkpointId, InputChannelInfo info, int startSeqNum, Buffer... data) {
+		LOG.debug("add input data, checkpoint id: {}, channel: {}, startSeqNum: {}, num buffers: {}",
+			checkpointId, info, startSeqNum, data == null ? 0 : data.length);
+		enqueue(write(checkpointId, info, checkBufferType(data)));
+	}
+
+	@Override
+	public void addOutputData(long checkpointId, ResultSubpartitionInfo info, int startSeqNum, Buffer... data) {
+		LOG.debug("add output data, checkpoint id: {}, channel: {}, startSeqNum: {}, num buffers: {}",
+			checkpointId, info, startSeqNum, data == null ? 0 : data.length);
+		enqueue(write(checkpointId, info, checkBufferType(data)));
+	}
+
+	@Override
+	public void finishInput(long checkpointId) {
+		LOG.debug("finish input data, checkpoint id: {}", checkpointId);
+		enqueue(completeInput(checkpointId));
+	}
+
+	@Override
+	public void finishOutput(long checkpointId) {
+		LOG.debug("finish output data, checkpoint id: {}", checkpointId);
+		enqueue(completeOutput(checkpointId));
+	}
+
+	@Override
+	public ChannelStateWriteResult getWriteResult(long checkpointId) {
+		LOG.debug("requested write result, checkpoint id: {}", checkpointId);
+		ChannelStateWriteResult result = results.remove(checkpointId);
+		Preconditions.checkArgument(result != null, "channel state write result not found for checkpoint id " + checkpointId);
+		return result;
+	}
+
+	public void open() {
+		Preconditions.checkState(isRunning);
+		asyncWriter.start();
+	}
+
+	@Override
+	public void close() {
+		isRunning = false;
+		results.clear();
+		asyncWriter.interrupt();
+		while (asyncWriter.isAlive()) {
+			try {
+				asyncWriter.join();
+			} catch (InterruptedException e) {
+				if (!asyncWriter.isAlive()) {
+					Thread.currentThread().interrupt();
+				}
+				LOG.debug("interrupted while waiting for the writer thread to die", e);
+			}
+		}
+		rethrow();
+	}
+
+	private void enqueue(ChannelStateWriteRequest request) {
+		rethrow();
+		Preconditions.checkState(isRunning);
+		try {
+			handover.put(request);
+		} catch (InterruptedException e) {
+			throw new RuntimeException("Interrupted while trying to add new handover request", e);
+		}
+	}
+
+	private void rethrow() {
+		if (thrown != null) {
+			throw new RuntimeException(thrown); // wrap to record current stack-trace
+		}
+	}
+
+	private static Buffer[] checkBufferType(Buffer... data) {
```

Comment:
Give some javadoc to explain why we do not support event data?

Reply:
The limitation itself is documented on the interface level.
The reasons behind it are too high-level for this class IMO.

Reply:
That is fine for me if it is documented in some places